### PR TITLE
account for batch padding when updating metrics

### PR DESCRIPTION
**Pull request by mikerossgithub**
Fixes executor_group.update_metric() to take account of batch padding. If a batch has padding, only the non padded data is used to update the metric. A unit test is included which demonstrates the fix.

### CODE REVIEW

File: python/mxnet/module/executor_group.py [hunk 1]
```diff
@@ -96,6 +96,19 @@ def _merge_multi_context(outputs, major_axis):
...
 
+def _slice_axis(arr, axis, islice):
+    """Slices array along axis"""
+    if axis == 0:
+        # slicing NDArray along axis 0 can avoid copying
+        return arr[islice]
+    elif axis > 0:
+        # pylint: disable=no-member
+        return nd.slice_axis(arr, axis=axis, begin=islice.start,
+                             end=islice.stop).as_in_context(arr.context)
```
Comment (piiswrong) [inline python/mxnet/module/executor_group.py]: don't need .as_in_context(arr.context)

also it's just nd.slice now

**Comment by jmerkow** [comment]
@piiswrong or @mikerossgithub, do you know if padding handled correctly when update (for parameters) is called?

File: python/mxnet/module/executor_group.py [hunk 2]
```diff
@@ -400,6 +414,12 @@ def forward(self, data_batch, is_train=None):
 
         """
         _load_data(data_batch, self.data_arrays, self.data_layouts)
+
+        if hasattr(data_batch, 'pad'):
```
Comment (jmerkow) [inline python/mxnet/module/executor_group.py]: You can replace this with `self.cur_batch_pad = getattr(data_batch, 'pad', None)`

**Comment by mikerossgithub** [comment]
@piiswrong @jmerkow thanks I made the changes suggested by both your comments and pushed.

**Comment by piiswrong** [comment]
@jmerkow Padding is not handled in updates. During training it's best to discard the last batch

**Comment by fhieber** [comment]
@piiswrong I think this change introduces a regression for symbols w/ SoftmaxOutput losses and perplexity metrics:
Before, one could apply SoftmaxOutput to 'flattened data': (batch_size * seq_len, classes) and compute mx.metric.Perplexity on these flattened predictions with a flattened one-hot label array (the metric also flattened the label shape).
After this change ( and #8003 ), the update_metric call requires the first axis to be the batch_size only (no collapsing between batch and seq_len allowed anymore). If this is changed accordingly in the symbol (namely using SoftmaxOutput with preserve_shape=True), returning (batch_size, seq_len, vocab_size) predictions, and using non-flattened one-hot labels, the Perplexity metric fails in update(), as labels are flattened internally, but not pred (in line metric.py::652).
So it seems that either all metrics should support multi-dimensional input, or the slicing from #8003 should be made less strict.

**Comment by piiswrong** [comment]
I didn't look at this PR carefully enough. Looks like it's causing a lot of problems and the logic also seems wrong on multiple devices.

I'll revert this

**Comment by mikerossgithub** [comment]
@piiswrong @fhieber I'm happy to look into modifying the fix. Can you provide further guidance or code snippets on what use cases are broken? For perplexity, would it be correct to unflatten the output data so that the padded portion can be sliced off?

**Comment by mikerossgithub** [comment]
@piiswrong Another approach I could implement would be to avoid slicing in `executor_group` and instead alter the metric.update methods to receive all padding-relevant information in **kwargs. Then it would be up to each metric to implement whether/how it handles batch padding. We could also give metrics which properly handle padding a common superclass. Would that be a better approach for mxnet?

**Comment by jmerkow** [comment]
@piiswrong I think that @mikerossgithub's approach could work pretty seamlessly.
Though I think padding needs to be addressed at multple
EvalMetric could very easily be changed to take batch information (such as padding) and the logic to remove this padding could be move there and used when necessary.  This could very easily be tied to a flag on EvalMetric to remove or not remove padding. Taken from [here](https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/metric.py#L87-L108) something like the following could be done:

```
def update_dict(self, label, pred, **batch_info):
   if self.remove_padding and batch_info.get('pad', 0): # self.remove_padding can be set on the class, or passed in.
      # do logic to remove padding from labels and press
      # batch_info can also include the batch axis etc if needed.


   # continue on with this function as usual.....
```

This way metrics could either have a default to remove padding (most metrics would do this), and those that need to leave labels and data shapes untouched would be able to do so.  It has the added advantage that if `update_dict` is called without the batch info you get the default behavior.

Im following this issue closely as it seems that metrics have been calculated __inaccurately__ when the batch size does not divide evenly into batches....
